src/utils/map_iter.h

```diff
@@ -21,14 +21,60 @@
 
 #include <vector>
 
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "src/common/logging/logging.h"
+
 namespace raksha::utils {
 
-template <typename T, typename U, typename F>
-std::vector<U> MapIter(const std::vector<T>& input, F f) {
-  std::vector<U> result;
+enum class CollisionPolicy {
+  kError,
+  kIgnore,
+};
+
+// A version of `Map` that iterates over a reference to a container to
+// populate a resultant container. The input container is taken as a const
+// reference here.
+template <typename ResultT, typename InputT, typename F>
+ResultT MapIterRef(const InputT& input, F f,
+                   CollisionPolicy collision_policy = CollisionPolicy::kError) {
+  ResultT result;
   result.reserve(input.size());
   for (auto& entity : input) {
-    result.push_back(f(entity));
+    uint64_t pre_insert_size = result.size();
+    result.insert(result.end(), f(entity));
+    // The usual way to check whether an insert succeeded or not would be to
+    // check the return value of `insert`. Unfortunately, this has different
+    // forms for `std::vector` versus maps or sets. Therefore, we just check
+    // that the size of the resultant collection has increased to ensure that
+    // the insert succeeded.
+    CHECK((collision_policy != CollisionPolicy::kError) ||
+          (pre_insert_size < result.size()))
+        << "Found collision when mapping a container.";
+  }
+  return result;
+}
+
+// A version of `Map` that iterates over a reference to a container to
+// populate a resultant container. The input container is taken as a value
+// here, which is better for if the input can be moved or is a temporary.
+template <typename ResultT, typename InputT, typename F>
+ResultT MapIterVal(InputT input, F f,
```

Comment:
Here's the big issue that I'm trying to deal with with the two implementations: if we were to only have a version that took the value type `InputT` as a parameter, we would incur a copy at the callsite whenever we only had a `const` reference as input. Does that matter? Since the implementation of `MapIter` basically copies all of the contents of the input container into the output container, maybe it doesn't in most cases.

One place I can see it mattering is if the container contains `std::unique_ptr`s. In that case, the copy will be illegal. But perhaps if we exposed a `MapIter` version that took a container by value and a second version that took iterators we would cover all of our bases. I think I'm going to try that.

Reply:
Wouldn't having an `iterator_range` solve the issue with copy vs move?

Reply:
You're right, actually, it would. OK, I'll go with that.

Reply:
What's the advantage of using a single `iterator_range` parameter versus two iterator parameters?

Reply:
Oh, I see: you can use the nicer syntax of a range-based for loop. Ok, makes sense.

Reply:
Ah, I don't think that is actually helpful here. If we use a range-based for loop, we have to decide within the for loop whether to write `auto current_element` or `auto &current_element`, which is awkward for smoothly handling regular and move iterators with the same code. If we use the form with iterators, however, we can write the for loop in terms of the iterators, call `f(*iter)` in the body, and therefore have the decision of whether to receive a value or a reference be made at the parameter of the lambda, which is supplied at the callsite.

Reply:
You're right, I hadn't considered this. Alright, returning a `vector` works for me.